core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java

```diff
@@ -0,0 +1,417 @@
+  private final Function<Map<String, String>, RESTClient> clientBuilder;
+  private RESTClient client = null;
+  private String catalogName = null;
+  private Map<String, String> properties = null;
+  private Object conf = null;
+  private FileIO io = null;
+
+  RESTCatalog(Function<Map<String, String>, RESTClient> clientBuilder) {
+    this.clientBuilder = clientBuilder;
+  }
+
+  @Override
+  public void initialize(String name, Map<String, String> props) {
+    this.client = clientBuilder.apply(props);
+    this.catalogName = name;
+    this.properties = ImmutableMap.copyOf(props);
+    String ioImpl = props.get(CatalogProperties.FILE_IO_IMPL);
+    this.io = CatalogUtil.loadFileIO(ioImpl != null ? ioImpl : ResolvingFileIO.class.getName(), props, conf);
+  }
+
+  @Override
+  public void setConf(Configuration newConf) {
+    this.conf = newConf;
+  }
+
+  @Override
+  public String name() {
+    return catalogName;
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    String ns = RESTUtil.urlEncode(namespace);
+    ListTablesResponse response = client
+        .get("v1/namespaces/" + ns + "/tables", ListTablesResponse.class, ErrorHandlers.namespaceErrorHandler());
+    return response.identifiers();
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    String tablePath = tablePath(identifier);
+    // TODO: support purge flagN
+    DropTableResponse response = client.delete(
+        tablePath, DropTableResponse.class, ErrorHandlers.tableErrorHandler());
+    return response.isDropped();
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+
+  }
+
+  private LoadTableResponse loadInternal(TableIdentifier identifier) {
+    String tablePath = tablePath(identifier);
+    return client.get(tablePath, LoadTableResponse.class, ErrorHandlers.tableErrorHandler());
+  }
+
+  @Override
+  public Table loadTable(TableIdentifier identifier) {
+    LoadTableResponse response = loadInternal(identifier);
+    Pair<RESTClient, FileIO> clients = tableClients(response.config());
+
+    return new BaseTable(
+        new RESTTableOperations(clients.first(), tablePath(identifier), clients.second(), response.tableMetadata()),
+        fullTableName(identifier));
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {
+    CreateNamespaceRequest request = CreateNamespaceRequest.builder()
+        .withNamespace(namespace)
+        .setProperties(metadata)
+        .build();
+
+    // for now, ignore the response because there is no way to return it
+    client.post("v1/namespaces", request, CreateNamespaceResponse.class, ErrorHandlers.namespaceErrorHandler());
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) throws NoSuchNamespaceException {
+    Preconditions.checkArgument(namespace.isEmpty(), "Cannot list namespaces under parent: %s", namespace);
```

Comment:
"Cannot list namespaces with a parent specified: %s passed" ?

The current message has an implication that perhaps another parent would work.